File: web/src/multiplayer/protocol.ts
// Wire types for the table service — mirrors server/src/protocol.rs.

import type { BetKind, FlipRequest, PlacedBet, RoundSnapshot, Side } from "../engine/types";
import type { TableTier } from "../tables";

/** Chairs at a table — mirrors the server's MAX_SEATS. */
export const MAX_SEATS = 7;

/** One seat's public face, shown to the whole table. */
export interface SeatView {
  id: number;
  name: string;
  bankroll: number;
  staked: number;
  /** This seat's staged bets — the same money `staked` totals. */
  bets: PlacedBet[];
  sitting_out: boolean;
  /** Declared ready to deal (requires a bet). Reset every coup. */
  ready: boolean;
  /** Sitting out, ready, or broke — the deal waits for everyone to decide. */
  decided: boolean;
  /** Bankroll can't cover the table minimum, so this seat can't bet at all. */
  broke?: boolean;
}

/** The player's snapshot plus everyone's seat summaries and squeeze rights. */
export type TableViewMsg = RoundSnapshot & {
  seats: SeatView[];
  player_squeezer: number | null;
  banker_squeezer: number | null;
};

export interface RoomInfo {
  id: string;
  tier: TableTier;
  seats: number;
  max_seats: number;
  /** Spectators at the rail (absent from a server that predates them). */
  watchers?: number;
}

export type ClientMsg =
  | { type: "list_rooms" }
  | { type: "create_room"; name: string; tier: TableTier; private: boolean }
  | { type: "join_room"; room: string; name: string }
  | { type: "rejoin"; room: string; token: string }
  /** Stand behind the seats: every push, no chair. Works on a full table. */
  | { type: "watch"; room: string }
  /** Keepalive while watching — a spectator otherwise never speaks. */
  | { type: "ping" }
  | { type: "leave" }
  | { type: "rename"; name: string }
  | { type: "bet"; kind: BetKind; amount: number }
  | { type: "sit_out" }
  | { type: "clear_bets" }
  /** Declare ready to deal (requires a bet); once everyone is, the coup deals. */
  | { type: "ready" }
  /** Take back a ready declaration. */
  | { type: "unready" }
  | { type: "deal" }
  | { type: "peek"; hand: Side; index: number }
  | { type: "reveal"; hand: Side; index: number }
  /** Ask the dealer to turn one/both house cards early (the squeezer only). */
  | { type: "dealer_flip"; count: FlipRequest }
  | { type: "settle" }
  | { type: "new_shoe" };

export type ServerMsg =
  | { type: "rooms"; rooms: RoomInfo[] }
  | { type: "announce"; message: string }
  | {
      type: "joined";
      room: string;
      player: number;
      tier: TableTier;
      view: TableViewMsg;
      proto?: number;
      /** Credential for reclaiming this seat, bankroll intact, after a drop. */
      token?: string;
    }
  | {
      type: "watching";
      room: string;
      tier: TableTier;
      view: TableViewMsg;
      proto?: number;
      watchers?: number;
    }
  | { type: "state"; view: TableViewMsg; watchers?: number }
  /** Stood up. `reason` when the server did it (the table closed under us). */
  | { type: "left"; reason?: string }
  | { type: "error"; message: string }
  | { type: "closed"; reason: string };

/**
 * Where the table service lives, if it is not the host serving this page.
 *
 * Set `VITE_WS_URL` at build time when the site and the table service are on
 * different hosts — a static site on GitHub Pages plus a server elsewhere. When
 * one host serves both (a VPS, or the server's own SPA_DIR), leave it unset and
 * the client uses `/ws` on its own origin, which needs no configuration and
 * cannot drift out of date.
 */
export function socketUrl(): string {
  const configured = import.meta.env.VITE_WS_URL as string | undefined;
  if (configured) return configured;
  // Same origin as the page. `wss:` follows `https:` so a secure page never
  // opens an insecure socket, which browsers block as mixed content anyway.
  const proto = location.protocol === "https:" ? "wss:" : "ws:";
  return `${proto}//${location.host}/ws`;
}

/**
 * The reconnect credential for a room, kept in sessionStorage: it should
 * survive a reload or a dropped socket, but not outlive the browser tab — a
 * stale token on a shared machine is a seat with someone's money on it.
 */
const SEAT_KEY = "baccarat.seat";

export function saveSeatToken(room: string, token: string): void {
  try {
    sessionStorage.setItem(SEAT_KEY, JSON.stringify({ room, token }));
  } catch {
    // private mode, or storage full — reconnect degrades to a fresh buy-in
  }
}

/** The stored token for `room`, if the last seat taken was at that table. */
export function loadSeatToken(room: string): string | null {
  try {
    const raw = sessionStorage.getItem(SEAT_KEY);
    if (!raw) return null;
    const parsed = JSON.parse(raw) as { room?: unknown; token?: unknown };
    if (parsed.room !== room) return null;
    return typeof parsed.token === "string" ? parsed.token : null;
  } catch {
    return null;
  }
}

/** Any stored seat, for reconnecting without knowing the room up front. */
export function loadSeat(): { room: string; token: string } | null {
  try {
    const raw = sessionStorage.getItem(SEAT_KEY);
    if (!raw) return null;
    const p = JSON.parse(raw) as { room?: unknown; token?: unknown };
    return typeof p.room === "string" && typeof p.token === "string"
      ? { room: p.room, token: p.token }
      : null;
  } catch {
    return null;
  }
}

export function clearSeatToken(): void {
  try {
    sessionStorage.removeItem(SEAT_KEY);
  } catch {
    /* nothing to do */
  }
}

/**
 * The table this tab is watching, so a dropped socket (or a reload) puts the
 * spectator back at the same rail. No credential is needed — watching is
 * open to anyone with the code — so this is just the code. Same lifetime as
 * the seat token: the tab, not the browser.
 */
const WATCH_KEY = "baccarat.watch";

export function saveWatchRoom(room: string): void {
  try {
    sessionStorage.setItem(WATCH_KEY, room);
  } catch {
    /* private mode, or storage full — a reconnect lands in the lobby */
  }
}

export function loadWatchRoom(): string | null {
  try {
    const room = sessionStorage.getItem(WATCH_KEY);
    return room && room.length > 0 ? room : null;
  } catch {
    return null;
  }
}

export function clearWatchRoom(): void {
  try {
    sessionStorage.removeItem(WATCH_KEY);
  } catch {
    /* nothing to do */
  }
}

File: server/src/protocol.rs
//! The wire protocol: JSON over WebSocket, reusing the engine's serializable
//! types so the web client's generated typings keep working.

use baccarat_engine::scoreboard::Side;
use baccarat_engine::session::BetKind;
use baccarat_engine::table::{FlipRequest, PlayerId, TableView};
use serde::{Deserialize, Serialize};

/// Bumped on any breaking wire change. Sent with `Joined` so a stale client
/// can tell "please refresh" apart from a generic bad message.
pub const PROTOCOL_VERSION: u32 = 1;

/// Stake tiers, mirrored from the web client's tables.ts.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Tier {
    Low,
    Mid,
    High,
}

impl Tier {
    /// (table_min, table_max, buy_in) in cents.
    pub fn stakes(self) -> (i64, i64, i64) {
        match self {
            Tier::Low => (100, 50_000, 50_000),
            Tier::Mid => (2_500, 500_000, 1_000_000),
            Tier::High => (50_000, 50_000_000, 25_000_000),
        }
    }
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMsg {
    ListRooms,
    CreateRoom { name: String, tier: Tier, private: bool },
    JoinRoom { room: String, name: String },
    /// Reclaim a seat kept warm after a drop. The token came from `Joined`.
    Rejoin { room: String, token: String },
    /// Stand behind the seats: every push the table gets, no chair, no chips.
    /// Works on a full table and, by code, on a private one.
    Watch { room: String },
    /// Keepalive. A spectator sends nothing else, and a silent connection is
    /// otherwise evicted as away (see IDLE_LIMIT).
    Ping,
    /// Stand up — from a seat or from the rail.
    Leave,
    /// Change the name the table sees for this seat. Sanitized like a join.
    Rename { name: String },
    Bet { kind: BetKind, amount: i64 },
    SitOut,
    ClearBets,
    Deal,
    Peek { hand: Side, index: usize },
    Reveal { hand: Side, index: usize },
    /// The squeezer asks the dealer to turn one or both of the house hand's
    /// cards before finishing their own — the high-limit courtesy.
    DealerFlip { count: FlipRequest },
    Settle,
    NewShoe,
}

#[derive(Debug, Clone, Serialize)]
pub struct RoomInfo {
    pub id: String,
    pub tier: Tier,
    pub seats: usize,
    pub max_seats: usize,
    /// Spectators at the rail.
    pub watchers: usize,
}

#[derive(Debug, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMsg {
    Rooms { rooms: Vec<RoomInfo> },
    /// The dealer's voice between flips ("Turning the Banker's cards…").
    Announce { message: String },
    Joined {
        room: String,
        player: PlayerId,
        tier: Tier,
        view: TableView,
        proto: u32,
        /// Bearer credential for reclaiming this exact seat, bankroll intact,
        /// if the socket drops. Sent only to its owner; never broadcast.
        token: String,
    },
    /// Standing at the rail: the spectator's view (see `Table::view_public`).
    /// No token — there is no seat to reclaim; a dropped watcher just watches
    /// again.
    Watching { room: String, tier: Tier, view: TableView, proto: u32, watchers: usize },
    State {
        view: TableView,
        /// Spectators at the rail, so the table can show who's watching.
        watchers: usize,
    },
    /// Stood up. `reason` when the server did it — the table closed under a
    /// spectator — so the client can say why in the lobby.
    Left {
        #[serde(skip_serializing_if = "Option::is_none")]
        reason: Option<String>,
    },
    Error { message: String },
    /// The server is closing this connection for a stated reason (e.g. the
    /// seat was given up after being away). The client shows `reason`.
    Closed { reason: String },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn client_messages_parse_from_json() {
        let m: ClientMsg = serde_json::from_str(
            r#"{"type":"create_room","name":"sab","tier":"mid","private":true}"#,
        )
        .unwrap();
        assert!(matches!(m, ClientMsg::CreateRoom { tier: Tier::Mid, private: true, .. }));

        let m: ClientMsg =
            serde_json::from_str(r#"{"type":"bet","kind":{"Main":"Player"},"amount":2500}"#)
                .unwrap();
        assert!(matches!(m, ClientMsg::Bet { amount: 2500, .. }));

        let m: ClientMsg =
            serde_json::from_str(r#"{"type":"peek","hand":"Banker","index":1}"#).unwrap();
        assert!(matches!(m, ClientMsg::Peek { index: 1, .. }));

        let m: ClientMsg =
            serde_json::from_str(r#"{"type":"dealer_flip","count":"Both"}"#).unwrap();
        assert!(matches!(m, ClientMsg::DealerFlip { count: FlipRequest::Both }));

        let m: ClientMsg = serde_json::from_str(r#"{"type":"rename","name":"alice"}"#).unwrap();
        assert!(matches!(m, ClientMsg::Rename { ref name } if name == "alice"));

        let m: ClientMsg = serde_json::from_str(r#"{"type":"watch","room":"ab12cd"}"#).unwrap();
        assert!(matches!(m, ClientMsg::Watch { ref room } if room == "ab12cd"));
        let m: ClientMsg = serde_json::from_str(r#"{"type":"ping"}"#).unwrap();
        assert!(matches!(m, ClientMsg::Ping));
    }

    #[test]
    fn a_plain_left_carries_no_reason_field() {
        // Older clients match `{"type":"left"}` exactly; a server-initiated
        // stand-up adds the reason, a voluntary one stays as it was.
        assert_eq!(serde_json::to_string(&ServerMsg::Left { reason: None }).unwrap(), r#"{"type":"left"}"#);
        assert_eq!(
            serde_json::to_string(&ServerMsg::Left { reason: Some("closed".into()) }).unwrap(),
            r#"{"type":"left","reason":"closed"}"#
        );
    }

    #[test]
    fn tier_stakes_match_the_web_tables() {
        assert_eq!(Tier::Low.stakes(), (100, 50_000, 50_000));
        assert_eq!(Tier::Mid.stakes(), (2_500, 500_000, 1_000_000));
        assert_eq!(Tier::High.stakes(), (50_000, 50_000_000, 25_000_000));
    }
}
